### PR TITLE
Fix flaky test AsyncDataCacheTest.ttl

### LINKED ISSUE #8189: AsyncDataCacheTest.ttl is flaky
### Bug description

[Expected behavior] and [actual behavior].5: [ RUN      ] AsyncDataCacheTest.ttl
5: I20231227 07:32:14.231534 28752 SsdFile.cpp:774] [SSDCA] Starting shard 0 without checkpoint
5: I20231227 07:32:14.231607 28752 SsdFile.cpp:774] [SSDCA] Starting shard 1 without checkpoint
5: I20231227 07:32:14.231676 28752 SsdFile.cpp:774] [SSDCA] Starting shard 2 without checkpoint
5: I20231227 07:32:14.231743 28752 SsdFile.cpp:774] [SSDCA] Starting shard 3 without checkpoint
5: I20231227 07:32:14.304281 31243 AsyncDataCache.cpp:534] [SSDCA] Limiting SSD save batch to 588 entries
5: I20231227 07:32:14.304381 31243 AsyncDataCache.cpp:534] [SSDCA] Limiting SSD save batch to 627 entries
5: I20231227 07:32:14.304404 31243 AsyncDataCache.cpp:534] [SSDCA] Limiting SSD save batch to 564 entries
5: I20231227 07:32:14.314743 31243 SsdFile.cpp:661] [SSDCA] Checkpointing shard 0, force: 0 bytesAfterCheckpoint: 1.72MB checkpointIntervalBytes: 1.60MB
5: I20231227 07:32:14.324473 31243 SsdCache.cpp:128] [SSDCA] Wrote 6MB, 0.0658458 MB/s
5: I20231227 07:32:14.339534 31264 AsyncDataCache.cpp:579] [CACHE] Removed 509 AsyncDataCache entries.
5: I20231227 07:32:14.342849 31264 AsyncDataCache.cpp:579] [CACHE] Removed 499 AsyncDataCache entries.
5: I20231227 07:32:14.346889 31264 AsyncDataCache.cpp:579] [CACHE] Removed 545 AsyncDataCache entries.
5: I20231227 07:32:14.348112 31263 AsyncDataCache.cpp:579] [CACHE] Removed 0 AsyncDataCache entries.
5: I20231227 07:32:14.348253 31263 AsyncDataCache.cpp:579] [CACHE] Removed 0 AsyncDataCache entries.
5: I20231227 07:32:14.348361 31263 AsyncDataCache.cpp:579] [CACHE] Removed 0 AsyncDataCache entries.
5: I20231227 07:32:14.349735 31264 AsyncDataCache.cpp:579] [CACHE] Removed 493 AsyncDataCache entries.
5: I20231227 07:32:14.349841 31263 AsyncDataCache.cpp:579] [CACHE] Removed 0 AsyncDataCache entries.
5: I20231227 07:32:14.350323 31263 SsdFile.cpp:586] [SSDCA] Removed 172 entries from /tmp/velox_test_4ioSnI/cache0. And erased 1 regions with 50% entries removed.
5: I20231227 07:32:14.350550 31263 SsdFile.cpp:586] [SSDCA] Removed 157 entries from /tmp/velox_test_4ioSnI/cache1. And erased 1 regions with 50% entries removed.
5: I20231227 07:32:14.350858 31263 SsdFile.cpp:586] [SSDCA] Removed 149 entries from /tmp/velox_test_4ioSnI/cache2. And erased 1 regions with 50% entries removed.
5: I20231227 07:32:14.350876 31264 CacheTTLController.cpp:68] Failed applying cache TTL of 98 seconds. Entries from 66 files are to be removed, while 0 files are retained
5: I20231227 07:32:14.351198 31263 SsdFile.cpp:586] [SSDCA] Removed 150 entries from /tmp/velox_test_4ioSnI/cache3. And erased 1 regions with 50% entries removed.
5: I20231227 07:32:14.351220 31263 CacheTTLController.cpp:68] Succeeded applying cache TTL of 98 seconds. Entries from 66 files are to be removed, while 0 files are retained
5: ../../velox/common/caching/tests/AsyncDataCacheTest.cpp:1146: Failure
5: Expected equality of these values:
5:   statsTtl.ssdStats->entriesAgedOut
5:     Which is: 628
5:   statsT1.ssdStats->entriesCached
5:     Which is: 172
5: W20231227 07:32:14.360149 28752 File.cpp:129] close failure in LocalReadFile destructor: -1, Bad file descriptor
5: W20231227 07:32:14.360194 28752 File.cpp:129] close failure in LocalReadFile destructor: -1, Bad file descriptor
5: W20231227 07:32:14.360208 28752 File.cpp:129] close failure in LocalReadFile destructor: -1, Bad file descriptor
5: W20231227 07:32:14.360221 28752 File.cpp:129] close failure in LocalReadFile destructor: -1, Bad file descriptor
5: I20231227 07:32:14.367921 28752 TempDirectoryPath.cpp:31] TempDirectoryPath:: removing all files from/tmp/velox_test_4ioSnI
5: [  FAILED  ] AsyncDataCacheTest.ttl (136 ms)
5: [----------] 10 tests from AsyncDataCacheTest (29659 ms total)

### System information

This happens on circle ci Linux build run: https://circleci.com/api/v1.1/project/github/facebookincubator/velox/285728/output/107/0?file=true&allocation-id=658bcac2aced6d10c464c307-0-build%2FABCDEFGH

### Relevant logs

```bash
https://circleci.com/api/v1.1/project/github/facebookincubator/velox/285728/output/107/0?file=true&allocation-id=658bcac2aced6d10c464c307-0-build%2FABCDEFGH
```

### PR DESCRIPTION
**Pull request by gggrace14**
Wait until SSD writes finish to collect SSD stats.

Fixed #8189

### CODE REVIEW

File: velox/common/caching/tests/AsyncDataCacheTest.cpp [hunk 1]
```diff
@@ -505,6 +505,9 @@ void AsyncDataCacheTest::loadNFiles(
   for (auto i = 0; i < numLoads; i++) {
     semaphore.acquire();
   }
+  while (cache_->ssdCache()->writeInProgress()) {
+    std::this_thread::sleep_for(std::chrono::milliseconds(100)); // NOLINT
+  }
 }
```
Comment (xiaoxmeng) [inline velox/common/caching/tests/AsyncDataCacheTest.cpp:512]: #ifndef NDEBUG
DEBUG_ONLY_TEST_F(AsyncDataCacheTest, ttl)

You can remove "#ifndef NDEBUG", DEBUG_ONLY_TEST_F only works on debug build.

File: velox/common/caching/tests/AsyncDataCacheTest.cpp [hunk 2]
```diff
@@ -505,6 +505,9 @@ void AsyncDataCacheTest::loadNFiles(
   for (auto i = 0; i < numLoads; i++) {
     semaphore.acquire();
   }
+  while (cache_->ssdCache()->writeInProgress()) {
```
Comment (xiaoxmeng) [inline velox/common/caching/tests/AsyncDataCacheTest.cpp:508]: Shall we put this right above the test code that check the TTL stats? Thanks!

**Review by xiaoxmeng — COMMENTED** [review]
@gggrace14 thanks for fix!

File: velox/common/caching/tests/AsyncDataCacheTest.cpp (hunk 1 above)
Comment (gggrace14) [reply velox/common/caching/tests/AsyncDataCacheTest.cpp]: Hi @xiaoxmeng  Revise. Can you take a look again?

File: velox/common/caching/tests/AsyncDataCacheTest.cpp (hunk 2 above)
Comment (gggrace14) [reply velox/common/caching/tests/AsyncDataCacheTest.cpp]: Revises

File: velox/common/caching/tests/AsyncDataCacheTest.cpp [hunk 3]
```diff
@@ -1129,21 +1128,25 @@ DEBUG_ONLY_TEST_F(AsyncDataCacheTest, ttl) {
 
   stt.setCurrentTestTimeSec(loadTime1);
   loadNFiles(filenames_.size() * 2 / 3, offsets);
+  while (cache_->ssdCache()->writeInProgress()) {
```
Comment (xiaoxmeng) [inline velox/common/caching/tests/AsyncDataCacheTest.cpp:1131]: Maybe put this into a test utility: waitForSSDWriteToFinish()? thanks!

**Review by xiaoxmeng — APPROVED** [review]
@gggrace14 LGTM. Thanks!

File: velox/common/caching/tests/AsyncDataCacheTest.cpp (hunk 3 above)
Comment (gggrace14) [reply velox/common/caching/tests/AsyncDataCacheTest.cpp]: Thanks, @xiaoxmeng , for the review! Revised according to above.

**Pull request closed without merging**